### ISSUE TITLE
Extension does not play well with Run-From-Zip deployment

### ISSUE THREAD

**Issue by smoelker** [sjkp/letsencrypt-siteextension#239]
Microsoft has introduced a new way to deploy your WebApps to Azure called [Run-From-Zip](https://github.com/Azure/app-service-announcements/issues/84) that will be out of preview at the end of Q3. This method basically allows you to mount a zip file to the wwwroot folder of your app at startup. This comes with some great benefits but also a big downside:  the wwwroot folder becomes read-only.

The Let's Encrypt site extension relies on writing the to the wwwroot folder (or whatever folder is specified in `letsencrypt:WebRootPath`) for serving the answer of the ACME challenge.

I've tried using the extension with Run-From-Zip deployment but the process of requesting a certificate fails. The Let's Encrypt error report tells me that the request to `https://my-site.ext/.well-known/acme-challenge/{challenge}` failed because of a HTTP 404 (Not Found).

I would expect this process to fail since the wwwroot folder is frozen (well, actually I would expect the process to fail at _writing_ the answer to the wwwroot folder but I might not completely understand the internals of the Run-From-Zip method).

My current workaround is to have the extension write the answers to a different folder (d:\home\data\letsencrypt\challenges) by setting the `letsencrypt:WebRootPath` and have the application serving any requests to `/.well-known/acme-challenge/{challenge}`.

A better solution would be if the extension did not rely on writing to the wwwroot folder for serving the answers but handled it internally by having a handler listening to the path `/.well-known/acme-challenge/{challenge}` instead.

**Comment by xt0rted** [comment]
I just updated one of my sites to use the run from package setup and the way I worked around the `.well-known` folder issue is like so (taken from https://github.com/projectkudu/kudu/issues/2849#issuecomment-426438325):

1. Create a folder called `LetsEncrypt` under `D:\home\site`
1. Create a folder called `.well-known` under `D:\home\site\LetsEncrypt`
1. In the site's *Application settings* create a new virtual path called `/.well-known` which points to `site\LetsEncrypt\.well-known`
1. Add an *App Setting* called `letsencrypt:WebRootPath` with a value of `D:\home\site\LetsEncrypt`
1. Once the site's using run from package uninstall and reinstall the Let's Encrypt Site extension, this looks to force the webjob  to be setup outside of the `wwwroot` folder

My site doesn't use the `.well-known` folder right now, but on my test site I did and had that working by:

1. Creating a virtual path called `/.well-known` which points to `site\wwwroot\wwwroot\.well-known` (or where ever your real folder is stored)
1. Creating a second virtual path called `/.well-known/acme-challenge` which points to `site\LetsEncrypt\.well-known\acme-challenge`

**Comment by Alexei-B** [comment]
I didn't want to use a virtual path as @xt0rted did, however, I managed to get this to work by:
1. Setting the `letsencrypt:WebRootPath` environment variable to a different location under the home directory.
2. Modifying the "letsencrypt" function to read from that directory.

Sample Code (.NET Standard):
```
        {
            log.LogInformation($"Processing Let's Encrypt Request for Code: {code}");

            var content = File.ReadAllText(@"D:\home\lets-encrypt\.well-known\acme-challenge\"+code);
            return new OkObjectResult(content);
        }
```

This approach worked just fine for me and it is much easier to automate without the extra virtual directory.

**Comment by alastairs** [comment]
@Alexei-B's approach worked for us too, after a _lot_ of frustration today. @sjkp Please could the wiki be updated with these instructions? 😄 I can't see any way on the Functions wiki page to contribute this change myself.

**Comment by sjkp** [comment]
@iyerusad a few comments, thanks for your feedback. 

%WEBSITE_HOSTNAME% is well supported, it just expands an environment variable. The WEBSITE_HOSTNAME environment variable is always present it is set by the app service. It is something newer that they support localhost. 

The route in the C# is not needed as it is implicit from the function and parameter name. 

I can't really change the default behavior of writing to wwwroot, as it would be a breaking change, unless the extension in some otherway can ensure that it work without the users configuring anything extra. I can't think of a way to do that atm. Suggestions are welcome :) 

Also please take a look at release 0.8.9 it now supports not writing the challenge file to disk, but instead writing it to blob-storage, that could maybe make life easier for those of you that use deploy from zip.

**Comment by danzel** [comment]
Looks like VSTS deploys default to deploy from zip now, the Kudu error message has changed too (pasting to make this issue easier to find in search)

Server Error in '/letsencrypt' Application.
Response status code does not indicate success: 409 (Conflict). 

And the stack trace starts with:
```
[HttpRequestException: Response status code does not indicate success: 409 (Conflict).]
   System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode() +224
   LetsEncrypt.Azure.Core.<PutFile>d__11.MoveNext() in D:\a\1\s\LetsEncrypt.SiteExtension.Core\KuduRestClient.cs:71
```

**Comment by sjkp** [comment]
@danzel thanks for letting me know. Guess I have to play around with deploy from zip, to make sure it works as smooth as possible.

**Comment by Justincale** [comment]
@xt0rted Thanks so much, this workes a treat after a few days of turmoil. !! :)

**Comment by tomaszzmuda** [comment]
In the version 0.9.3 right now
![image](https://user-images.githubusercontent.com/13612984/52780911-97dc7a00-304b-11e9-8d14-2679988da924.png)

**Comment by sjkp** [comment]
@tomaszzmuda nothing have been done to fix the problem from the extension side of things.

**Comment by Misaka-0x447f** [comment]
azure pipeline user, meet the same problem.

**Comment by Misaka-0x447f** [comment]
plus, microsoft maybe going to work on it? https://feedback.azure.com/forums/170024-additional-services/suggestions/16957756-add-integration-with-let-s-encrypt

**Comment by sjkp** [comment]
Have you heard anything new from Microsoft? It sounds like radio silence to me still.

**Comment by Misaka-0x447f** [comment]
Emmmmm nope. It's one of the top feedback but opened 2 yrs ago.

**Comment by sjkp** [comment]
Fixed in 0.9.4

**Comment by mmartain** [comment]
I have the same problem. Does not work out of the box...
